### PR TITLE
Add implementation plan for documenting the API media properties

### LINKED ISSUE #1996: Implementation Plan: Clearly document all media properties in catalog in API & Frontend

### PR DESCRIPTION
**Pull request by dhruvkb**
## Fixes

Fixes #1996 by @obulat

## Description

This PR adds the implementation plan for documenting the media properties in the API. The IP uses introspection as the primary approach along with a documentation generation plan similar to that for the catalog.

## Reviewers

I have requested reviews from @obulat and @krysal. The reason for the selection is because @obulat was involved in the IP and implementation for the corresponding project at the catalog layer and because IIRC @krysal has done prior work in the area of media properties documentation.

## Testing Instructions

Please read the IP and comment your opinions. You can also see the implementation of this plan at #3966.

## Decision-making process

This discussion is following the Openverse decision-making process. Information about this process can be found [on the Openverse documentation site](https://docs.openverse.org/meta/decision_making/index.html). Requested reviewers or participants will be following this process. If you are being asked to give input on a specific detail, you do not need to familiarise yourself with the process and follow it.

### Current round

This discussion is currently in the **Clarification round**. The deadline for review of this round is 5 April 2024.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by krysal** [comment]
I'm reviewing this today. In the meantime, @dhruvkb, could you add a note about the revision round, like in #3650, for example?

File: documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+- The `max_length` data-point is not captured because it is automatically
+  covered under the `db_type` data-point.
+- `is_pk`, `needs_unique`, `allows_null` and `allows_blank` are presented
+  collectively as "constraints" although not technically DB-level constraints.
+```
+
+The information extracted from this introspection will be stored in memory for
+the next stage.
+
+### Merge with additional notes
```
Comment (krysal) [inline documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md:117]: This section is appreciated; surely something will come up here 😄

File: documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+### Generate documentation page
+
+With all information in-memory about the media properties, the output will be
+written to a page by the management command. This requires some restructuring in
+the pages created for the catalog media properties documentation.
+
+Similar to the catalog documentation, the output document will include the
+`preamble.md` file from `api/api/docs/media_properties/preamble.md`. This will
+be followed by the Markdown generated from the media properties as a table.
```
Comment (krysal) [inline documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md:157]: Could you include the intended destination of this file? Like in the [catalog documentation](https://docs.openverse.org/_preview/3965/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.html#media-properties-md).

File: documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+## Workstreams
+
+The entire plan can be implemented in one stream. A lot of components of the
+plan are shared with the catalog media properties documentation but code reuse
+between the two projects will be limited as they follow different idioms.
+
+The project assumes that the implementation of the catalog documentation project
+will be completed before this project begins, and includes some reorganisation
+of the output of that project. If that is not the case, the reorganisation can
+be incorporated into the ongoing work on that project.
```
Comment (krysal) [inline documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md:175]: I see you were waiting for the catalog workstream to be completed. However, I believe the file's final destination should be defined in this IP before starting the code work. It's okay if it's changed later for any (justified) reason.

File: documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+Although we have used utility scripts for the catalog documentation, for the API
+we will be done using a management command as that is more idiomatic and gives
+us access to the Django ORM models.
+
+## Future improvements
+
+This implementation plan does not document the schema for Django's `JSONField`
+(`jsonb`) in much detail. Further iterations can implement this an as a
+potential improvement.
```
Comment (krysal) [inline documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md:206]: Sounds reasonable to me 👍

**Review by krysal — COMMENTED** [review]
@dhruvkb This is looking like a solid document! I don't anticipate more questions or comments than the ones I left here.

File: documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md (hunk 4 above)
Comment (dhruvkb) [reply documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md:206]: For now, it is possible to document some shape of the file in the additional notes section but we can automate/formalise it in the future using some introspection within Python or from the database.

File: documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md (hunk 3 above)
Comment (dhruvkb) [reply documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md]: I've added the new hierarchy and the file destination in 416e492.

File: documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md (hunk 2 above)
Comment (dhruvkb) [reply documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md]: Yes! Added in 416e492.

File: documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md [hunk 5]
```diff
@@ -0,0 +1,239 @@
+# 2024-03-25 Implementation Plan: Document all media properties in the API
+
+**Author**: @dhruvkb
+
+## Reviewers
+
+<!-- Choose two people at your discretion who make sense to review this based on their existing expertise. Check in to make sure folks aren't currently reviewing more than one other proposal or RFC. -->
+
+- [ ] @obulat
+- [ ] @krysal
```
Comment (krysal) [inline documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md:10]:
```suggestion
- [x] @krysal
```

**Review by krysal — APPROVED** [review]
Thanks for updating the details, and writing this comprehensive proposal! Everything looks good to me, so shortcuting the rounds, I approve it 👍

File: documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+
+On its own the command will be insufficient to write the documentation page to
+the documentation project as it will run inside the Django context and will not
+have access to the filesystem. This management command will write the
+documentation to the API project root.
+
+We will invoke this command from a `just` recipe which will move the file to its
+destination, lint it and also function as the CI test which returns a non-zero
+exit code when the file is modified in the process. This recipe will be named
+`gen-media-props`.
```
Comment (obulat) [inline documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md:61]: We could name the script more generally, like `generate-docs` with argument of `media-props` (as in the catalog PR). If we ever have other generated docs in the API, then it would be easier to update.
Just a suggestion to unify the naming across the stack.

File: documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+
+Additional notes or hardcoded information may be extra information about a field
+such as known inconsistencies or special cases. This information, if documented
+using a consistent format, can be used to augment the information extracted in
+the previous step.
+
+We will document this information in the docstring of the model classes from
+where it can be accessed using `inspect.getdoc`. To ensure that these notes are
+co-located with the fields' code, we will parse docstrings of all ancestors and
+mixins of a model class.
```
Comment (obulat) [inline documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md:127]: Will this work with `help_text` that is located in a separate file and is added inside `__init__`, as the `source` ones?
https://github.com/WordPress/openverse/blob/12d7162fb9a9c44a019ea329d87f573447792bc8/api/api/serializers/media_serializers.py#L310-L311

File: documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+
+The final page will be written to a Markdown file at the path
+`documentation/meta/media_properties/api.md`. To include this Markdown file an
+index page will be required at `documentation/meta/media_properties/index.md`,
+which will itself be included in the `documentation/meta/index.md` file.
+
+```{note}
+If the catalog media properties documentation is already created at
+`documentation/meta/media_properties.md` by the time this implementation plan is
+implementation, it will need to be moved one-level deeper to
```
Comment (obulat) [inline documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md:163]:
```suggestion
implemented, it will need to be moved one-level deeper to
```

File: documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+We will be using Django's own introspection capabilities. No additional
+libraries other than the already installed Django packages will be needed.
+
+The documentation generated will be added as a page to documentation site
+without needed any new packages.
+
+## Alternatives
+
+The approach for the catalog uses SQL files. Unlike that, we will be using
+Django's introspection capabilities are quite powerful and allow for document
```
Comment (obulat) [inline documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md:215]:
```suggestion
Django's introspection capabilities which are quite powerful and allow for document
```

**Review by obulat — COMMENTED** [review]
It's great that Django has all of the capabilities to allow for documenting the properties! Love the plan and the linked PR with documentation.

I would love to see more detail on the `Future improvements` point on `jsonb` fields, especially after the incident caused by #3930. Do you think this should be done in a separate project, or as a separate part of this project? Would we add separate serializers to allow us to use introspection, or is there another way available to auto-generate documentation?

I wouldn't consider the "Fully document media properties" project completed without documenting the JSONFields because they are the ones that can cause problems in the future.

File: documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md (hunk 7 above)
Comment (dhruvkb) [reply documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md:127]: These help texts are taken from the model and not from the serializer. That's because for a specific model a number of serializers may be defined that can have different `help_text`s as per the requirements of the endpoint.

So for source, we show this

https://github.com/WordPress/openverse/blob/12d7162fb9a9c44a019ea329d87f573447792bc8/api/api/models/media.py#L56-L58

File: documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md (hunk 9 above)
Comment (dhruvkb) [reply documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md]: Good catch, fixed in d0661a9.

File: documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md (hunk 8 above)
Comment (dhruvkb) [reply documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md]: Good catch, fixed in d0661a9.

File: documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md (hunk 6 above)
Comment (dhruvkb) [reply documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md]: That makes sense, I didn't consider the convention that was established in the catalog by #3620. Updated in 166bb42.

File: documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md (hunk 7 above)
Comment (dhruvkb) [reply documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md:128]: You can visit https://docs.openverse.org/_preview/3966/meta/media_properties/api.html to see the the generated docs output.

File: documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md [hunk 10]
```diff
@@ -0,0 +1,265 @@
+# 2024-03-25 Implementation Plan: Document all media properties in the API
+
+**Author**: @dhruvkb
+
+## Reviewers
+
+<!-- Choose two people at your discretion who make sense to review this based on their existing expertise. Check in to make sure folks aren't currently reviewing more than one other proposal or RFC. -->
+
+- [ ] @obulat
```
Comment (obulat) [inline documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md:9]:
```suggestion
- [x] @obulat
```

**Review by obulat — COMMENTED** [review]
Thank you for adding information on the json fields and possible solutions.
The plan looks good, and I'm short-circuiting this discussion to approve it.

**Review by obulat — APPROVED** [review]
